In [30]:
# using python 3 to better handle unicode emojis
import emoji
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import statsmodels.formula.api as smf

pd.set_option('display.max_columns',500)
pd.set_option('display.max_rows',500)

%matplotlib inline

# Goal: 
Using   the   training   data   in    ages_train.csv ,   build   a   statistical   model   to   predict   the
age   of   users   in    ages_test.csv,     based   on   any   of   the   information   provided   (profiles, friend   networks,   and   mentions).

# Load data

In [2]:
# load json files: 
ap = pd.read_json('assignment_package/age_profiles.json')
at = pd.read_json('assignment_package/age_tweets.json')
# fp = pd.read_json('assignment_package/friend_profiles.json')
# mp = pd.read_json('assignment_package/mention_profiles.json')

In [3]:
# load csv files:
ages_train = pd.read_csv('assignment_package/ages_train.csv')
mentions = pd.read_csv('assignment_package/mentions.csv')
friends = pd.read_csv('assignment_package/friends.csv')

# Take a look at data

In [4]:
mentions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17695 entries, 0 to 17694
Data columns (total 3 columns):
ID                 17695 non-null int64
MentionedID        17695 non-null int64
MentionedHandle    17695 non-null object
dtypes: int64(2), object(1)
memory usage: 414.8+ KB


In [5]:
mentions.head()

ID  MentionedID MentionedHandle
0  612815277    798747864    AlenaBacalia
1  612815277    354590068      hfholden34
2  612815277    401111769   MarilynYousif
3  612815277     35581752      heeeeyerin
4  612815277   1591956090      DogeTheDog

In [6]:
friends.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46609 entries, 0 to 46608
Data columns (total 2 columns):
ID          46609 non-null int64
FriendID    46609 non-null int64
dtypes: int64(2)
memory usage: 728.3 KB


In [7]:
friends.head()

ID   FriendID
0  612815277  574147452
1  612815277   36767299
2  612815277   29374598
3  612815277   67634814
4  612815277   30008146

In [8]:
ages_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1711 entries, 0 to 1710
Data columns (total 2 columns):
ID     1711 non-null int64
Age    1711 non-null int64
dtypes: int64(2)
memory usage: 26.8 KB


In [9]:
ages_train.head()

ID  Age
0   23605717   21
1  781084760   19
2  102262916   21
3  413710015   21
4  584304916   90

In [10]:
ap.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2410 entries, 0 to 2409
Data columns (total 41 columns):
contributors_enabled                  2410 non-null bool
created_at                            2410 non-null datetime64[ns]
default_profile                       2410 non-null bool
default_profile_image                 2410 non-null bool
description                           2410 non-null object
entities                              2410 non-null object
favourites_count                      2410 non-null int64
follow_request_sent                   2410 non-null bool
followers_count                       2410 non-null int64
following                             2410 non-null bool
friends_count                         2410 non-null int64
geo_enabled                           2410 non-null bool
id                                    2410 non-null int64
id_str                                2410 non-null int64
is_translation_enabled                2410 non-null bool
is_translator      

In [11]:
ap.head()

contributors_enabled          created_at  default_profile  \
0                 False 2012-06-19 17:54:58            False   
1                 False 2012-02-12 22:36:17            False   
2                 False 2011-03-11 21:44:56            False   
3                 False 2009-12-17 20:08:55            False   
4                 False 2011-03-24 21:14:53            False   

   default_profile_image                                        description  \
0                  False             You'll never find another one like me.   
1                  False  If you like me, i probably like you more. {IG:...   
2                  False          Success doesn’t come to you…you go to it.   
3                  False  NEVER care about what a person says that doesn...   
4                  False  these niggas some haters & i made myself so ea...   

                        entities  favourites_count  follow_request_sent  \
0  {'description': {'urls': []}}              7931                False   
1  {'description': {'urls': []}}              1141                False   
2  {'description': {'urls': []}}               174                False   
3  {'description': {'urls': []}}               303                False   
4  {'description': {'urls': []}}                52                False   

   followers_count  following  friends_count  geo_enabled         id  \
0              274      False            264         True  612815277   
1              481      False            527         True  490791079   
2              878      False            953        False  264416315   
3              742      False            701        False   97515432   
4              729      False            688        False  271608323   

      id_str  is_translation_enabled  is_translator lang  listed_count  \
0  612815277                   False          False   en             1   
1  490791079                   False          False   en             1   
2  264416315                   False          False   en             2   
3   97515432                   False          False   en            30   
4  271608323                   False          False   en             0   

                location          name  notifications  \
0                               Greggy          False   
1                 9⃣3⃣7⃣        Guff✌️          False   
2            Wonderland   Starqueria.⚓          False   
3  •Carolina South• 843•   . keeleesha          False   
4       Seven•five•Seven   killakee :)          False   

  profile_background_color                       profile_background_image_url  \
0                   131516  http://abs.twimg.com/images/themes/theme14/bg.gif   
1                   1A1B1F   http://abs.twimg.com/images/themes/theme9/bg.gif   
2                   FCEBB6  http://pbs.twimg.com/profile_background_images...   
3                   131516  http://pbs.twimg.com/profile_background_images...   
4                   0A0A0A  http://pbs.twimg.com/profile_background_images...   

                  profile_background_image_url_https  profile_background_tile  \
0  https://abs.twimg.com/images/themes/theme14/bg...                     True   
1  https://abs.twimg.com/images/themes/theme9/bg.gif                    False   
2  https://pbs.twimg.com/profile_background_image...                     True   
3  https://pbs.twimg.com/profile_background_image...                     True   
4  https://pbs.twimg.com/profile_background_image...                     True   

                                  profile_banner_url  \
0  https://pbs.twimg.com/profile_banners/61281527...   
1  https://pbs.twimg.com/profile_banners/49079107...   
2  https://pbs.twimg.com/profile_banners/26441631...   
3                                                NaN   
4  https://pbs.twimg.com/profile_banners/27160832...   

                                   profile_image_url  \
0  http://pbs.twimg.com/profile_images/4776662828...   
1  http://pbs.twimg.com/profile_images/509722754

In [12]:
at.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28138 entries, 0 to 28137
Data columns (total 25 columns):
contributors                 0 non-null float64
coordinates                  1298 non-null object
created_at                   28138 non-null datetime64[ns]
entities                     28138 non-null object
favorite_count               28138 non-null int64
favorited                    28138 non-null bool
geo                          1298 non-null object
id                           28138 non-null int64
id_str                       28138 non-null int64
in_reply_to_screen_name      5299 non-null object
in_reply_to_status_id        4839 non-null float64
in_reply_to_status_id_str    4839 non-null float64
in_reply_to_user_id          5299 non-null float64
in_reply_to_user_id_str      5299 non-null float64
lang                         28138 non-null object
metadata                     28138 non-null object
place                        1341 non-null object
possibly_sensitive          

In [13]:
at.head()

contributors coordinates          created_at  \
0           NaN        None 2014-09-26 05:55:45   
1           NaN        None 2014-09-26 05:58:17   
2           NaN        None 2014-09-26 05:59:47   
3           NaN        None 2014-09-26 06:00:07   
4           NaN        None 2014-09-26 08:09:57   

                                            entities  favorite_count  \
0  {'hashtags': [], 'urls': [], 'symbols': [], 'u...               0   
1  {'hashtags': [], 'urls': [], 'symbols': [], 'u...               0   
2  {'hashtags': [], 'urls': [], 'symbols': [], 'u...               1   
3  {'hashtags': [], 'urls': [], 'symbols': [], 'u...               1   
4  {'hashtags': [], 'urls': [], 'symbols': [], 'u...               1   

   favorited   geo                  id              id_str  \
0      False  None  515379174454747136  515379174454747136   
1      False  None  515379813670854656  515379813670854656   
2      False  None  515380190361288704  515380190361288704   
3      False  None  515380273005854721  515380273005854720   
4      False  None  515412946608852992  515412946608852992   

  in_reply_to_screen_name  in_reply_to_status_id  in_reply_to_status_id_str  \
0                    None                    NaN                        NaN   
1                    None                    NaN                        NaN   
2                    None                    NaN                        NaN   
3                    None                    NaN                        NaN   
4                    None                    NaN                        NaN   

   in_reply_to_user_id  in_reply_to_user_id_str lang  \
0                  NaN                      NaN   en   
1                  NaN                      NaN   en   
2                  NaN                      NaN   en   
3                  NaN                      NaN   en   
4                  NaN                      NaN   en   

                                            metadata place  \
0  {'result_type': 'recent', 'iso_language_code':...  None   
1  {'result_type': 'recent', 'iso_language_code':...  None   
2  {'result_type': 'recent', 'iso_language_code':...  None   
3  {'result_type': 'recent', 'iso_language_code':...  None   
4  {'result_type': 'recent', 'iso_language_code':...  None   

   possibly_sensitive  retweet_count  retweeted  \
0                 NaN              7      False   
1                 NaN             80      False   
2                 NaN              1      False   
3                 NaN              0      False   
4                 NaN              0      False   

                                    retweeted_status  \
0  {'in_reply_to_status_id': None, 'coordinates':...   
1  {'in_reply_to_status_id': None, 'coordinates':...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/android" ...   

                                                text  truncated  \
0  RT @AlenaBacalia: i over think and second gues...      False   
1  RT @hfholden34: I've wasted too much time on t...      False   
2                                  I'm already gone.      False   
3  I'm already something to someone that I don't ...      False   
4  I don't know what I'm doing with my life, but ...      False   

                                                user  
0  {'follow_request_sent': False, 'profile_image_...  
1  {'follow_request_sent': False, 'profile_image_...  
2  {'follow_request_sent': False, 'profile_image_...  
3  {'follow_request_sent': False, 'profile_image_...  
4  {'follow_request_sent': False, 'profile_ima

# First Iteration: Use all numeric variables to fit a linear regression model

In [23]:
def extract_emojis(s):
    return ''.join(c for c in str(s) if c in emoji.UNICODE_EMOJI)

def age_profiles_prep(age_profiles, ages_train):
    '''Process age_profiles for modeling'''
    ap_new = pd.concat([age_profiles,
                        pd.DataFrame(age_profiles['status'].apply(pd.Series) \
                        .rename(columns = lambda x: 'status_' + str(x)))], axis=1)
    ap_new['status_emoji'] = ap_new['status_text'].map(lambda x: extract_emojis(x))
    ap_new['emoji_cnt'] = ap_new['status_emoji'].map(lambda x: len(x))
    ap_new = ap_new.set_index('id')
    ages_train = ages_train.set_index('ID')
    age_joined = ap_new.join(ages_train, how='inner')
    return age_joined

In [24]:
age_joined = age_profiles_prep(ap, ages_train)

//anaconda/envs/capstone/lib/python3.6/site-packages/pandas/core/indexes/api.py:43: RuntimeWarning: '>' not supported between instances of 'int' and 'str', sort order is undefined for incomparable objects
  union = _union_indexes(indexes)
//anaconda/envs/capstone/lib/python3.6/site-packages/pandas/core/indexes/api.py:77: RuntimeWarning: '>' not supported between instances of 'int' and 'str', sort order is undefined for incomparable objects
  result = result.union(other)


#### ages_train dataset has 1711 user IDs, age_profiles has 2410 users IDs, after inner join with age_profiles, return 1688 user IDs.

In [26]:
age_joined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1688 entries, 612815277 to 2385386499
Data columns (total 67 columns):
contributors_enabled                  1688 non-null bool
created_at                            1688 non-null datetime64[ns]
default_profile                       1688 non-null bool
default_profile_image                 1688 non-null bool
description                           1688 non-null object
entities                              1688 non-null object
favourites_count                      1688 non-null int64
follow_request_sent                   1688 non-null bool
followers_count                       1688 non-null int64
following                             1688 non-null bool
friends_count                         1688 non-null int64
geo_enabled                           1688 non-null bool
id_str                                1688 non-null int64
is_translation_enabled                1688 non-null bool
is_translator                         1688 non-null bool
lang  

In [25]:
age_joined.head()

contributors_enabled          created_at  default_profile  \
612815277                 False 2012-06-19 17:54:58            False   
490791079                 False 2012-02-12 22:36:17            False   
264416315                 False 2011-03-11 21:44:56            False   
97515432                  False 2009-12-17 20:08:55            False   
224078321                 False 2010-12-08 02:50:16            False   

           default_profile_image  \
612815277                  False   
490791079                  False   
264416315                  False   
97515432                   False   
224078321                  False   

                                                 description  \
612815277             You'll never find another one like me.   
490791079  If you like me, i probably like you more. {IG:...   
264416315          Success doesn’t come to you…you go to it.   
97515432   NEVER care about what a person says that doesn...   
224078321                                          STOP PREE   

                                entities  favourites_count  \
612815277  {'description': {'urls': []}}              7931   
490791079  {'description': {'urls': []}}              1141   
264416315  {'description': {'urls': []}}               174   
97515432   {'description': {'urls': []}}               303   
224078321  {'description': {'urls': []}}              1188   

           follow_request_sent  followers_count  following  friends_count  \
612815277                False              274      False            264   
490791079                False              481      False            527   
264416315                False              878      False            953   
97515432                 False              742      False            701   
224078321                False             1030      False           1006   

           geo_enabled     id_str  is_translation_enabled  is_translator lang  \
612815277         True  612815277                   False          False   en   
490791079         True  490791079                   False          False   en   
264416315        False  264416315                   False          False   en   
97515432         False   97515432                   False          False   en   
224078321         True  224078321                   False          False   en   

           listed_count               location          name  notifications  \
612815277             1                               Greggy          False   
490791079             1                 9⃣3⃣7⃣        Guff✌️          False   
264416315             2            Wonderland   Starqueria.⚓          False   
97515432             30  •Carolina South• 843•   . keeleesha          False   
224078321             3                     ME         King.          False   

          profile_background_color  \
612815277                   131516   
490791079                   1A1B1F   
264416315                   FCEBB6   
97515432                    131516   
224078321                   642D8B   

                                profile_background_image_url  \
612815277  http://abs.twimg.com/images/themes/theme14/bg.gif   
490791079   http://abs.twimg.com/images/themes/theme9/bg.gif   
264416315  http://pbs.twimg.com/profile_background_images...   
97515432   http://pbs.twimg.com/profile_background_images...   
224078321  http://pbs.twimg.com/profile_background_images...   

                          profile_background_image_url_https  \
612815277  https://abs.twimg.com/images/themes/theme14/bg...   
490791079  https://abs.twimg.com/images/themes/theme9/bg.gif   
264416315  https://pbs.twimg.com/profile_background_image...   
97515432   https://pbs.twimg.com/profile_background_image...   
224078321  https://pbs.twimg.com/profile_background_image...   

           profile_background_tile  \
612815277                     True   
490791079                    False   
264416315                     True   
97515432                   